docs/workers.md

```diff
@@ -232,6 +232,7 @@ information.
     ^/_matrix/client/(api/v1|r0|v3|unstable)/joined_rooms$
     ^/_matrix/client/v1/rooms/.*/timestamp_to_event$
     ^/_matrix/client/(api/v1|r0|v3|unstable)/search$
+    ^/_matrix/client/(r0|v3|unstable)/user/.*/filter(/|$)
```

Comment:
(Just starting a thread somewhere).

The REST servlet to add a filter:

1. Checks the filter JSON against what's in the database and if an exact match is found it returns the same ID (i.e. you cannot have duplicate filters w/ different IDs).
2. Otherwise it, finds the maximum current filter ID and increments by 1.
3. Inserts the new filter.

I *think* this change might break step 1 -- if we have two requests to insert a filter, they can both race and return no known filter in step 1, then both insert the filter with different IDs?

Reply:
(Note the only index is UNIQUE across `(user_id, filter_id)`.)

Reply:
You were almost right (I think) that there was a phantom write here. But funnily enough, step 1 won't be broken because step 2 will fix it. The reason I say this is:

* we run transactions at the `REPEATABLE READ` isolation level by default and do not override the default for this txn
* For your race condition to hold, Step 1 for both transactions needs to happen concurrently so neither sees the other's filter.
* Both transactions will therefore see the same maximum filter ID in step 2 — because of REPEATABLE READ and step (1)
* Both transactions will attempt to insert a filter with the same ID in step 3. One of them won't be able to.

When trying in `psql`, the second INSERT blocks until the first has committed, but then gives a unique key constraint violation. I don't think this is one of the errors we retry a transaction for, so maybe there is a problem here after all.

```
syn7=*> INSERT INTO user_filters (user_id, filter_id, filter_json) VALUES ('@user:syn7', 1, '{}');
ERROR:  duplicate key value violates unique constraint "user_filters_unique"
DETAIL:  Key (user_id, filter_id)=(@user:syn7, 1) already exists.
```

Reply:
Ah, yes good call with the isolation level we're using. 👍 

I think we *don't* retry that, but I'm not sure. I guess using a sequence here would have been better? (Would that have solved our issues?)

Reply:
(to be fair, this exact same issue already existed if the user made two requests at once!)

Reply:
Yes using a SEQUENCE or a SERIAL field would have saved this effort. I guess the original author really liked the idea of a user's filters starting at 1 and then incrementing by 1, rather than having the same space of IDs shared by all users.

Reply:
Another way of thinking of this would have been randomly generating them, like we do in a few other spots. (Incrementing IDs are usually a API smell.)

Reply:
(what's wrong with incrementing IDs?)

Reply:
> (what's wrong with incrementing IDs?)

If they're scoped by user it is likely fine. If they're *not* then it can (potentially) leak information about *other* users (or your service as a whole). It also makes it easier to guess IDs.